cdci_data_analysis/analysis/drupal_helper.py

```diff
@@ -52,39 +55,81 @@ def execute_drupal_request(url,
-                       f"try to solve it as soon as possible")
-        if sentry_client is not None:
-            sentry_client.capture('raven.events.Message',
-                                  message=f'exception when performing a request to the product gallery: {repr(e)}')
-        else:
-            logger.warning("sentry not used")
-
-        raise InternalError('issue when performing a request to the product gallery',
-                         status_code=500,
-                         payload={'error_message': str(e)})
+    n_tries_left = n_max_tries
+    while True:
+        try:
+            if method == 'get':
+                if params is None:
+                    params = {}
+                params['_format'] = request_format
+                res = requests.get(url,
+                                   params={**params},
+                                   headers=headers)
+
+            elif method == 'post':
+                if data is None:
+                    data = {}
+                if params is None:
+                    params = {}
+                params['_format'] = request_format
+                res = requests.post(url,
+                                    params={**params},
+                                    data=data,
+                                    files=files,
+                                    headers=headers
+                                    )
+            else:
+                raise NotImplementedError
+            if res.status_code == 403:
+                try:
+                    response_json = res.json()
+                    # a 403 has been noticed to be returned in two different cases:
+                    # * for not-valid token
+                    # * not-completed request
+                    error_msg = response_json['message']
+                except json.decoder.JSONDecodeError:
+                    error_msg = res.text
+                raise RequestNotAuthorized(error_msg)
+
+            if res.status_code == 404:
+                logger.warning(f"there seems to be some problem in completing a request to the product gallery,"
+                               " this might be due to a an error in the url, "
+                               "or because we are attempting to load a page that no longer exists, "
+                               "please check it and try to issue again the request")
+                raise InternalError('issue when performing a request to the product gallery',
+                                    status_code=500,
+                                    payload={'error_message': str(e)})
+
+            return res
+
+        except (ConnectionError,
+                RequestNotAuthorized,
+                requests.exceptions.ConnectionError,
+                requests.exceptions.Timeout) as e:
+            n_tries_left -= 1
+            if n_tries_left > 0:
+                if n_tries_left < round(n_max_tries / 2):
```

Comment:
I have a small suggestion which I found very useful in such cases. There is a better way to do it and easy one. Easy one is better here.

We can make a global (module-scope) counter of these issues. Global state is dangerous, but it's applicable here, if just for logging.
Variables `total_n_successful_post_requests`, `total_n_post_request_retries`.
Increment them accordinly, and in the log message add "average post retries per request since dispatcher start: total_n_post_request_retries/total_n_successful_post_requests".
You could then also add an extra warning `when n_max_tries - n_tries_left  > total_n_post_request_retries/total_n_successful_post_requests`

What people often do is to report such metric e.g. in last hour, as self-health check. But it's maybe unnecessarily complex.

Reply:
Thanks, very nice suggestion!